Add transition parameters to Tween

KoBeWi opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/7802
tbh this change makes the code more complex than it's probably worth it. I don't think there is better way though; I didn't want it to affect performance.

TODO:
- ~~figure out `amplitude`, `period` for BACK equations~~ It doesn't use them?
- ~~figure out `overshoot`, `amplitude`, `period` for ELASTIC equations~~ It's `period` and `damping` now.
- TRANS_SPRING might also get parameters, as it's similar to elastic
- maybe other transitions might get parameters too?
- ~~add compat methods~~

I'll need help with the equations. I did `overshoot` for TRANS_BACK, because it was obvious. I'd appreciate if someone could just drop me implementations for other equations >_> (I have no idea about these numbers)

Here's BACK with default overshoot and 5

https://github.com/godotengine/godot/assets/2223172/43f1c52d-80df-4399-a59b-8152df5c9e24

KoBeWi committed:
Add transition parameters to Tween

Qcaria commented:
In my opinion only the overshoot parameter makes sense in the BACK equations, as it is a single bump (so the frequency is fixed) and amplitude and overshot would do the same. 

On the other hand, for the elastic equations there are two fixed parameters that have the role of frequency and amplitude (again overshoot in this context means the same as amplitude to me). 

The amplitude can be controlled by changing for example the 10 in `float a = c * pow(2, 10 * t);`. The frequency  could be controlled by changing the 0.3 in `float p = d * 0.3f;` which corresponds more or less  to a period (so the smaller it is the higher the frequency). 

I attach a couple of images to show the effect of modifying these two parameters.
Changing the frequency parameter:
<img width="472" alt="Captura de pantalla 2023-09-23 a las 10 55 27" src="https://github.com/godotengine/godot/assets/38281560/46826b3d-63b3-46d1-9228-39199ec91ecd">

Changing the amplitude parameter:
<img width="472" alt="Captura de pantalla 2023-09-23 a las 10 56 15" src="https://github.com/godotengine/godot/assets/38281560/0faec785-5ab7-4ca6-a082-f089c993cb70">

This is my first comment, I hope to not be messing up the style too much.

KoBeWi commented:
Thanks, I pushed the parameters for elastic transition. I renamed amplitude to damping, because it seems to have reverse effect.

Here's elastic with different parameters:

https://github.com/godotengine/godot/assets/2223172/60619534-2127-4ab0-8943-b512664ee027

**1** is default, **2** is period 0.1, **3** is amplitude 5
The parameters have slightly different effect depending on ease (EASE_OUT doesn't even use damping).

Seems like the third parameter is now unused, so I guess I will be removing it.

KoBeWi commented:
Ok I removed the third parameter. I also tried to de-hardcode the parameter count. While I can't do much about the arguments, I made arrays use a constant and also changed easing equations to use a macro (the method definitions are super-repetitive).

The nice thing about the new parameters is that, as long as their count remains constant, we can tweak other easings without breaking compatibility at all. So if TRANS_SPRING or any other transitions were to use the parameters, it can be added at any point.

akien-mga commented on scene/animation/easing_equations.h:
```diff
@@ -55,44 +55,46 @@
  * SOFTWARE.
  */
 
+#define EASING_FUNC(type) static real_t type(real_t t, real_t b, real_t c, real_t d, real_t p1, real_t p2)
```
Remember to `#undef` it at the end of the file, especially as it's in a header.

TokageItLab commented:
This looks good, but would it be too many arguments to take into account what I commented on in https://github.com/godotengine/godot/pull/82306#discussion_r1485824451?

KoBeWi commented:
@TokageItLab I could add another method similar to `interpolate_variant()` that takes a struct (with the intention that you make the struct and then reuse it between calls), though RefCounted structs are kind of pain to write. Maybe this is something to be improved later, I think `interpolate_variant()` isn't commonly used. RefCounted struct wouldn't be compatible with the new structs once they get addd.

As for #82306, I'd just put the custom interpolator at the end. There isn't any argument limit in functions, it only affects convenience.

KoBeWi commented:
CI fails on compatibility, not sure how to handle it. There was no `.expected` file for 4.3.

akien-mga commented:
> CI fails on compatibility, not sure how to handle it. There was no `.expected` file for 4.3.

The file of `4.2-stable.expected` - that's changes between 4.2-stable and HEAD.

TokageItLab reviewed (dismissed):
Functionally, it makes sense. Nevertheless, I leave it to the product team to discuss whether this is the correct way to pass additional arguments.

akien-mga commented on scene/animation/tween.h:
```diff
@@ -134,6 +137,13 @@ class Tween : public RefCounted {
 	static void _bind_methods();
 
+#ifndef DISABLE_DEPRECATED
+	Ref<Tween> _set_trans_compat_82155(Tween::TransitionType p_trans);
```
@KoBeWi You missed renames in the header, which is why it failed CI.

akien-mga commented:
Looks like using `inf` as default value for parameters breaks godot-cpp, might need a fix there.

AThousandShips commented on doc/classes/PropertyTweener.xml:
```diff
@@ -120,5 +120,13 @@
+			<param index="0" name="param1" type="float" />
+			<param index="1" name="param2" type="float" default="inf" />
+			<description>
+				Sets parameters of the transition. They are used only in specific transitions, check [enum Tween.TransitionType] for details.
```
```suggestion
				Sets the parameters of the transition. They are used only in specific transitions, check [enum Tween.TransitionType] for details.
```

AThousandShips commented on doc/classes/Tween.xml:
```diff
@@ -294,6 +297,14 @@
+			<param index="0" name="param1" type="float" />
+			<param index="1" name="param2" type="float" default="inf" />
+			<description>
+				Sets parameters of the transition. They are used only in specific transitions, check [enum TransitionType] for details.
```
```suggestion
				Sets the parameters of the transition. They are used only in specific transitions, check [enum TransitionType] for details.
```
